lib/Dialect/Calyx/Transforms/RemoveCombGroups.cpp

```diff
@@ -0,0 +1,291 @@
+  return newOp;
+}
+
+using CombResRegMapping = DenseMap<Value, RegisterOp>;
+
+struct RemoveCombGroupsPattern : public OpRewritePattern<calyx::CombGroupOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  RemoveCombGroupsPattern(MLIRContext *ctx, CombResRegMapping *mapping)
+      : OpRewritePattern(ctx), combResRegMapping(mapping) {}
+
+  LogicalResult matchAndRewrite(calyx::CombGroupOp combGroup,
+                                PatternRewriter &rewriter) const override {
+
+    auto component = combGroup->getParentOfType<ComponentOp>();
+    auto group = rewriter.replaceOpWithNewOp<calyx::GroupOp>(
+        combGroup, combGroup.getName());
+    rewriter.mergeBlocks(combGroup.getBodyBlock(), group.getBodyBlock());
+
+    // Determine which cell results are read from the control schedule.
+    SetVector<Operation *> cellsAssigned;
+    for (auto op : group.getOps<calyx::AssignOp>()) {
+      auto defOp = dyn_cast<CellInterface>(op.getDest().getDefiningOp());
+      assert(defOp && "expected some assignment to a cell");
+      cellsAssigned.insert(defOp);
+    }
+
+    rewriter.setInsertionPointToStart(group.getBodyBlock());
+    auto oneConstant = rewriter.create<hw::ConstantOp>(
+        group.getLoc(), APInt(1, 1, /*unsigned=*/true));
+
+    // Maintain the set of cell results which have already been assigned to
+    // its register within this group.
+    SetVector<Value> alreadyAssignedResults;
+
+    // Collect register done signals. These are needed for generating the
+    // GroupDone signal.
+    SetVector<Value> registerDoneSigs;
+
+    // 1. Iterate over the cells assigned within the combinational group.
+    // 2. For any use of a cell result within the controls schedule.
+    // 3.  Ensure that the cell result has a register.
+    // 4.  Ensure that the cell result has been written to its register in this
+    //     group.
+    // We do not replace uses of the combinational results now, since the
+    // following code relies on a checking cell result value use in the
+    // control schedule, which needs to remain even when two combinational
+    // groups assign to the same cell.
+    for (auto cellOp : cellsAssigned) {
+      auto cell = dyn_cast<CellInterface>(cellOp);
+      for (auto combRes : cell.getOutputPorts()) {
+        for (auto &use : llvm::make_early_inc_range(combRes.getUses())) {
+          if (use.getOwner()->getParentOfType<calyx::ControlOp>()) {
+            auto combResReg = combResRegMapping->find(combRes);
+            if (combResReg == combResRegMapping->end()) {
+              // First time a registered variant of this result is needed.
+              auto reg = createReg(component, rewriter, combRes.getLoc(),
+                                   cell.instanceName(),
+                                   combRes.getType().getIntOrFloatBitWidth());
+              combResRegMapping->insert({combRes, reg});
+              combResReg = combResRegMapping->find(combRes);
```

Comment:
I think you might be able to use the return value of `insert` instead of immediately calling `find`.